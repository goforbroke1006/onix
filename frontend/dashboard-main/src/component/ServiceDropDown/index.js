import React from "react";

class ServiceDropDown extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            items: [],
            value: "",
        }
    }

    componentDidMount() {
        this.loadServices();
    }

    onChange = (event) => {
        let serviceName = event.target.value;

        console.log(serviceName);

        this.setState({value: serviceName});

        if (this.props.onChange)
            this.props.onChange(serviceName);
    }

    render() {
        return (
            <div>
                <select onChange={this.onChange} value={this.state.value}>
                    <option>----</option>
                    {this.state.items.map((object, index) => {
                        return (
                            <option key={"service-select-option-" + index}
                                    value={object.title}>
                                {object.title}
                            </option>
                        )
                    })}
                </select>
            </div>
        )
    }

     loadServices = () => {
        let baseUrl = process.env.REACT_APP_API_DASHBOARD_MAIN_BASE_ADDR ?? 'http://127.0.0.1:8082/api/dashboard-main';
        fetch(baseUrl + "/service")
            .then(response => response.json())
            .then(data => this.setState({items: data}))
    }
}

export default ServiceDropDown;