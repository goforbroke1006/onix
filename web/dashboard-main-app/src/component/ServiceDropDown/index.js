import React from "react";
import PropTypes from "prop-types";
import {Select} from "antd";

const {Option} = Select;

export default class ServiceDropDown extends React.Component {
    static propTypes = {
        onChange: PropTypes.func,
        provider: PropTypes.object,
    };

    constructor(props) {
        super(props);
        this.state = {
            items: [],
            value: "",
        }
    }

    componentDidMount() {
        this.loadServices();
    }

    onChange = (serviceName) => {
        console.log(serviceName);

        this.setState({value: serviceName});

        if (this.props.onChange)
            this.props.onChange(serviceName);
    }

    render() {
        let emptyOption;
        if (this.state.items.length === 0) {
            emptyOption = (<Option value={""}>-- no data --</Option>)
        } else {
            emptyOption = (<Option value={""}>-- none --</Option>)
        }

        return (
            <Select defaultValue={this.state.value} style={{width: 600}} onChange={this.onChange}>
                {emptyOption}
                {this.state.items.map((object, index) => {
                    return (
                        <Option key={"service-select-option-" + index}
                                value={object.title}>{object.title}</Option>
                    )
                })}
            </Select>
        )
    }

    loadServices = () => {
        this.props.provider.loadServices()
            .then(data => this.setState({items: data}))
    }
}
