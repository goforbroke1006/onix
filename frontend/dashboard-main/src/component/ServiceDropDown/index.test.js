import React from "react";
import {mount, shallow} from "enzyme";

import ServiceDropDown from "./index";

describe('<ServiceDropDown /> rendering', () => {
    it('renders correctly', () => {
        const client = new MockDashboardMainApiClient();
        const wrapper = mount(<ServiceDropDown provider={client}/>);

        fakeServiceListPromise.then(() => {
            wrapper.update();
            expect(wrapper).toMatchSnapshot();
        });
    });

    it('onChange callback works', () => {
        const expectedValue = "foo/backend";
        let actualValue = null;

        let onChangeSpy = function (serviceName) {
            actualValue = serviceName;
        }

        const wrapper = shallow(<ServiceDropDown provider={new MockDashboardMainApiClient()} onChange={onChangeSpy}/>);
        wrapper.find('select').at(0).simulate('change', {
            target: {value: 'foo/backend', name: 'foo/backend'}
        });

        expect(actualValue).toEqual(expectedValue);
    });
});

const fakeServiceListPromise = new Promise((resolve, reject) => {
    const fakeServicesList = [
        {title: "acme/backend"},
        {title: "foo/backend"},
        {title: "bar/backend"},
    ];

    resolve(fakeServicesList)
});

class MockDashboardMainApiClient {
    loadServices() {
        return fakeServiceListPromise;
    }
}