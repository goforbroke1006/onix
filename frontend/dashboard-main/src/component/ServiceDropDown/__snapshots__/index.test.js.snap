// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ServiceDropDown /> rendering renders correctly 1`] = `
<ServiceDropDown
  provider={MockDashboardMainApiClient {}}
>
  <div>
    <select
      onChange={[Function]}
      value=""
    >
      <option>
        ----
      </option>
      <option
        key="service-select-option-0"
        value="acme/backend"
      >
        acme/backend
      </option>
      <option
        key="service-select-option-1"
        value="foo/backend"
      >
        foo/backend
      </option>
      <option
        key="service-select-option-2"
        value="bar/backend"
      >
        bar/backend
      </option>
    </select>
  </div>
</ServiceDropDown>
`;
